Add interpolate-style overloads to aten::upsample* ops

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37177 Update interpolate to use new upsample overloads
* **#37176 Add interpolate-style overloads to aten::upsample* ops**
* #37175 Add support for float[]? arguments in native_functions.yaml



The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

dreiss committed:
Add interpolate-style overloads to aten::upsample* ops

The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

Note that this initial revision only updates upsample_nearest2d
(9 C++ functions).  I want to get agreement on the implementation
before I do the other 50.

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

[ghstack-poisoned]

dreiss commented:
@lara-hdr , is this the sort of change you were thinking about in #27176 ?

lara-hdr commented:
> @lara-hdr , is this the sort of change you were thinking about in #27176 ?

Yes, this looks great! Thanks!

lara-hdr commented on tools/autograd/derivatives.yaml:
```diff
@@ -1183,6 +1183,9 @@
 - name: upsample_nearest3d(Tensor self, int[3] output_size, float? scales_d=None, float? scales_h=None, float? scales_w=None) -> Tensor
   self: upsample_nearest3d_backward(grad, output_size, self.sizes(), scales_d, scales_h, scales_w)
 
+- name: upsample_nearest2d.vec(Tensor input, int[]? output_size, float[]? scale_factors) -> Tensor
```
Since "int[]?" does not have the length of output_size anymore either,  the signature of all versions of upsample (1d-2d-3d) are now the same.
The plan is to simplify this further to have a single function for all upsample cases?
If so would that be done in this PR as well? if not what is the reason not to?

dreiss replied:
Very good point.  One possible downside would be increased code size on mobile because we currently do selective builds at operator granularity, and bundling them together would mean that someone who just wants to use upsample_nearest2d would also get the code for 1d and 3d.  We have a plan to mitigate this, but I'm not sure when we'll have time to implement it.  I'll measure the code size impact and report back.

lara-hdr commented:
@neginraoof for FYI

dreiss commented:
@eellison , I'm looking at just merging nearest{1,2,3}d into one op and *linear{1,2,3}d into one op.  You mentioned that reasons not to might be:

> - has a greater risk of dealing with tricky BC breakage

I don't really understand.  Can you elaborate?

> - will have a much higher build size in mobile when just the selective build thing works 

Not for nearest, since they all just delegate to one implementation anyway.  Linear looks like a marginal effect on size, and that can be mitigated in the future.

> - carries much more semantic meaning so is harder to export / analyze / maintain an implementation that is not ATen (Any graph backend, NNC compilation guard elimination)

Do you think this is a major issue if we only fold the dimensionality and still have separate ops for seperate modes?

> - All of the dispatching and python ops get compiled away when run in TorchScript anyway, so I guess that it would be (marginally) slower

This is true for tracing, but not for scripting, I think.  Even in tracing, it should be at most a few ns per op.


The benefit is primarily reducing the amount of wrapper code to simplify maintenance.  Thoughts?

eellison commented:
@dreiss those comments were about making `torch.nn.functional.interpolate` one big op, not necessarily about merging nearest, linear, etc.

dreiss commented:
Okay, I spent a few hours on the unification, and I'm not convinced it's a win for maintainability.  It allows some supporting code to be unified, but also requires a lot of function prototypes to be copy/pasted.  I'm going to move ahead with just keeping all 7 upsampling modes separate.  We can always unify them later and just have `upsample_nearest2d` and `upsample_nearest3d` dispatch to the same function.

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

Note that this initial revision only updates upsample_nearest2d
(9 C++ functions).  I want to get agreement on the implementation
before I do the other 50.

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss commented:
This is probably the final version of this change, if anyone wants to take another look.

ailzhang commented:
cc: @dlibenzi @JackCaoG  This PR will require a change on pytorch/xla codegen

dlibenzi commented:
> cc: @dlibenzi @JackCaoG This PR will require a change on pytorch/xla codegen

Taking a look.

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

dreiss committed:
Update on "Add interpolate-style overloads to aten::upsample* ops"




The non-deprecated user-facing interface to these ops (F.interpolate)
has a good interface: output size and scale are both specified as
a scalar or list, and exactly one must be present.  These aten ops
have an older, clunkier interface where output size is required and
scales are specified as separate optional scalars per dimension.

This change adds new overloads to the aten ops that match the interface
of interpolate.  The plan is to eventually remove the old overloads,
resulting in roughly net-zero code added.  I also believe it is possible
to push this interface down further, eliminating multiple optional<double>
arguments, and simplifying the implementations.

The rollout plan is to land this, wait for a reasonable interval for
forwards-compatibility (maybe 1 week?), land the change that updates
interpolate to call these overloads, wait for a reasonable interval
for backwards-compatibility (maybe 6 months?), then remove the old
overloads.

This diff does not add the `.out` variants of the ops because they
are not currently accessible through any user-facing API.


Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989/)

Differential Revision: [D21209989](https://our.internmc.facebook.com/intern/diff/D21209989)

[ghstack-poisoned]

artcyb commented:
🤣
